portable/GCC/ARM_CM4F/port.c

```diff
@@ -543,23 +530,44 @@ void xPortSysTickHandler( void )
 		to be unsuspended then abandon the low power entry. */
 		if( eTaskConfirmSleepModeStatus() == eAbortSleep )
 		{
-			/* Restart from whatever is left in the count register to complete
-			this tick period. */
-			portNVIC_SYSTICK_LOAD_REG = portNVIC_SYSTICK_CURRENT_VALUE_REG;
-
-			/* Restart SysTick. */
-			portNVIC_SYSTICK_CTRL_REG |= portNVIC_SYSTICK_ENABLE_BIT;
-
-			/* Reset the reload register to the value required for normal tick
-			periods. */
-			portNVIC_SYSTICK_LOAD_REG = ulTimerCountsForOneTick - 1UL;
-
 			/* Re-enable interrupts - see comments above the cpsid instruction()
 			above. */
 			__asm volatile( "cpsie i" ::: "memory" );
 		}
 		else
 		{
+			/* Stop the SysTick momentarily.  The time the SysTick is stopped for
+			is accounted for as best it can be, but using the tickless mode will
+			inevitably result in some tiny drift of the time maintained by the
+			kernel with respect to calendar time. */
+			portNVIC_SYSTICK_CTRL_REG = ( portNVIC_SYSTICK_CLK_BIT_SETTING | portNVIC_SYSTICK_INT_BIT );
+
+			/* Calculate the reload value required to wait xExpectedIdleTime
+			tick periods.  -1 is used because this code normally executes part
+			way through the first tick period.  But if the SysTick IRQ is now
+			pending, then clear the IRQ, suppressing the first tick, and correct
+			the reload value to reflect that the second tick period is already
+			underway.  The expected idle time is always at least two ticks. */
+			ulReloadValue = portNVIC_SYSTICK_CURRENT_VALUE_REG + ( ulTimerCountsForOneTick * (xExpectedIdleTime - 1UL) );
+			if( ( portNVIC_ICSR_REG & portNVIC_PEND_SYSTICK_SET_BIT ) != 0 )
+			{
+				portNVIC_ICSR_REG = portNVIC_PEND_SYSTICK_CLEAR_BIT;
+				
+				/* Skip the correction described above if SysTick happened to
+				stop on zero.  In that case, there are still
+				ulTimerCountsForOneTick ticks left in the second tick period,
+				not zero, so the value in portNVIC_SYSTICK_CURRENT_VALUE_REG
+				already includes the correction normally done here. */
+				if( portNVIC_SYSTICK_CURRENT_VALUE_REG != 0 )

```

Comment:
Another concern here about configurations that use the "other" SysTick clock.  If the other clock is very slow compared to the core clock, then the SysTick could be zero here *without* the IRQ asserted.  If the SysTick were zero when the idle task calculated xExpectedIdleTime for the final time, then potentially it could still be zero here.  The code proposed here does not properly handle that case because it checks SysTick for zero only if the IRQ is asserted.  Does FreeRTOS not place any restrictions on this "other" clock?  The other clock seems to be MCU specific.  I think EFM32 uses 32kHz for the other clock.  It wouldn't be too hard to support these very slow SysTick clocks, but at some runtime cost to all configurations.